feat(dqlite): Implement NamespaceNotifyWatcher;

ycliuhw opened the pull request:
This PR implements the NamespaceNotifyWatcher for watching changes in a namespace.
This is a similar implementation to ValueWatcher but without the need to provide an expected change value. Any time events matching the change mask occur in the namespace, a notification is emitted.


## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

It's not been used yet, so just run unit tests.

## Documentation changes

No

## Links

**Jira card:** [JUJU-6250](https://warthogs.atlassian.net/browse/JUJU-6250)



[JUJU-6250]: https://warthogs.atlassian.net/browse/JUJU-6250?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

SimonRichardson commented:
Can you not accomplish the same thing as the following diff?

```diff
diff --git a/core/watcher/eventsource/value.go b/core/watcher/eventsource/value.go
index 3f58fe56f8..2444545586 100644
--- a/core/watcher/eventsource/value.go
+++ b/core/watcher/eventsource/value.go
@@ -19,10 +19,11 @@ import (
 type ValueWatcher struct {
 	*BaseWatcher
 
-	out         chan struct{}
-	namespace   string
-	changeValue string
-	changeMask  changestream.ChangeType
+	out        chan struct{}
+	namespace  string
+	changeMask changestream.ChangeType
+
+	filter func(changestream.ChangeEvent) bool
 
 	mapper Mapper
 }
@@ -35,9 +36,11 @@ func NewValueWatcher(base *BaseWatcher, namespace, changeValue string, changeMas
 		BaseWatcher: base,
 		out:         make(chan struct{}),
 		namespace:   namespace,
-		changeValue: changeValue,
-		changeMask:  changeMask,
-		mapper:      defaultMapper,
+		filter: func(ce changestream.ChangeEvent) bool {
+			return ce.Changed() == changeValue
+		},
+		changeMask: changeMask,
+		mapper:     defaultMapper,
 	}
 
 	w.tomb.Go(w.loop)
@@ -52,9 +55,47 @@ func NewValueMapperWatcher(base *BaseWatcher, namespace, changeValue string, cha
 		BaseWatcher: base,
 		out:         make(chan struct{}),
 		namespace:   namespace,
-		changeValue: changeValue,
-		changeMask:  changeMask,
-		mapper:      mapper,
+		filter: func(ce changestream.ChangeEvent) bool {
+			return ce.Changed() == changeValue
+		},
+		changeMask: changeMask,
+		mapper:     mapper,
+	}
+
+	w.tomb.Go(w.loop)
+	return w
+}
+
+// NewNamespaceNotifyWatcher returns a new watcher that receives changes from the input
+// base watcher's db/queue when changes in the namespace occur.
+func NewNamespaceNotifyWatcher(base *BaseWatcher, namespace string, changeMask changestream.ChangeType) *ValueWatcher {
+	w := &ValueWatcher{
+		BaseWatcher: base,
+		out:         make(chan struct{}),
+		namespace:   namespace,
+		filter: func(changestream.ChangeEvent) bool {
+			return true
+		},
+		changeMask: changeMask,
+		mapper:     defaultMapper,
+	}
+
+	w.tomb.Go(w.loop)
+	return w
+}
+
+// NewNamespaceNotifyMapperWatcher returns a new watcher that receives changes from
+// the input base watcher's db/queue when changes in the namespace occur.
+func NewNamespaceNotifyMapperWatcher(base *BaseWatcher, namespace string, changeMask changestream.ChangeType, mapper Mapper) *ValueWatcher {
+	w := &ValueWatcher{
+		BaseWatcher: base,
+		out:         make(chan struct{}),
+		namespace:   namespace,
+		filter: func(changestream.ChangeEvent) bool {
+			return true
+		},
+		changeMask: changeMask,
+		mapper:     mapper,
 	}
 
 	w.tomb.Go(w.loop)
@@ -70,9 +111,7 @@ func (w *ValueWatcher) Changes() <-chan struct{} {
 func (w *ValueWatcher) loop() error {
 	defer close(w.out)
 
-	opt := changestream.FilteredNamespace(w.namespace, w.changeMask, func(e changestream.ChangeEvent) bool {
-		return e.Changed() == w.changeValue
-	})
+	opt := changestream.FilteredNamespace(w.namespace, w.changeMask, w.filter)
 	subscription, err := w.watchableDB.Subscribe(opt)
 	if err != nil {
 		return errors.Annotatef(err, "subscribing to namespace %q", w.namespace)
@@ -101,7 +140,7 @@ func (w *ValueWatcher) loop() error {
 			return ErrSubscriptionClosed
 		case changes, ok := <-in:
 			if !ok {
-				w.logger.Debugf("change channel closed for %q; terminating watcher for %q", w.namespace, w.changeValue)
+				w.logger.Debugf("change channel closed for %q; terminating watcher", w.namespace)
 				return nil
 			}
 
@@ -129,7 +168,7 @@ func (w *ValueWatcher) drainInitialEvent(in <-chan []changestream.ChangeEvent) {
 	select {
 	case _, ok := <-in:
 		if !ok {
-			w.logger.Debugf("change channel closed for %q; terminating watcher for %q", w.namespace, w.changeValue)
+			w.logger.Debugf("change channel closed for %q; terminating watcher", w.namespace)
 			return
 		}
 	default:
```

ycliuhw commented:
> Can you not accomplish the same thing as the following diff?
> 
> ```diff
> diff --git a/core/watcher/eventsource/value.go b/core/watcher/eventsource/value.go
> index 3f58fe56f8..2444545586 100644
> --- a/core/watcher/eventsource/value.go
> +++ b/core/watcher/eventsource/value.go
> @@ -19,10 +19,11 @@ import (
>  type ValueWatcher struct {
>  	*BaseWatcher
>  
> -	out         chan struct{}
> -	namespace   string
> -	changeValue string
> -	changeMask  changestream.ChangeType
> +	out        chan struct{}
> +	namespace  string
> +	changeMask changestream.ChangeType
> +
> +	filter func(changestream.ChangeEvent) bool
>  
>  	mapper Mapper
>  }
> @@ -35,9 +36,11 @@ func NewValueWatcher(base *BaseWatcher, namespace, changeValue string, changeMas
>  		BaseWatcher: base,
>  		out:         make(chan struct{}),
>  		namespace:   namespace,
> -		changeValue: changeValue,
> -		changeMask:  changeMask,
> -		mapper:      defaultMapper,
> +		filter: func(ce changestream.ChangeEvent) bool {
> +			return ce.Changed() == changeValue
> +		},
> +		changeMask: changeMask,
> +		mapper:     defaultMapper,
>  	}
>  
>  	w.tomb.Go(w.loop)
> @@ -52,9 +55,47 @@ func NewValueMapperWatcher(base *BaseWatcher, namespace, changeValue string, cha
>  		BaseWatcher: base,
>  		out:         make(chan struct{}),
>  		namespace:   namespace,
> -		changeValue: changeValue,
> -		changeMask:  changeMask,
> -		mapper:      mapper,
> +		filter: func(ce changestream.ChangeEvent) bool {
> +			return ce.Changed() == changeValue
> +		},
> +		changeMask: changeMask,
> +		mapper:     mapper,
> +	}
> +
> +	w.tomb.Go(w.loop)
> +	return w
> +}
> +
> +// NewNamespaceNotifyWatcher returns a new watcher that receives changes from the input
> +// base watcher's db/queue when changes in the namespace occur.
> +func NewNamespaceNotifyWatcher(base *BaseWatcher, namespace string, changeMask changestream.ChangeType) *ValueWatcher {
> +	w := &ValueWatcher{
> +		BaseWatcher: base,
> +		out:         make(chan struct{}),
> +		namespace:   namespace,
> +		filter: func(changestream.ChangeEvent) bool {
> +			return true
> +		},
> +		changeMask: changeMask,
> +		mapper:     defaultMapper,
> +	}
> +
> +	w.tomb.Go(w.loop)
> +	return w
> +}
> +
> +// NewNamespaceNotifyMapperWatcher returns a new watcher that receives changes from
> +// the input base watcher's db/queue when changes in the namespace occur.
> +func NewNamespaceNotifyMapperWatcher(base *BaseWatcher, namespace string, changeMask changestream.ChangeType, mapper Mapper) *ValueWatcher {
> +	w := &ValueWatcher{
> +		BaseWatcher: base,
> +		out:         make(chan struct{}),
> +		namespace:   namespace,
> +		filter: func(changestream.ChangeEvent) bool {
> +			return true
> +		},
> +		changeMask: changeMask,
> +		mapper:     mapper,
>  	}
>  
>  	w.tomb.Go(w.loop)
> @@ -70,9 +111,7 @@ func (w *ValueWatcher) Changes() <-chan struct{} {
>  func (w *ValueWatcher) loop() error {
>  	defer close(w.out)
>  
> -	opt := changestream.FilteredNamespace(w.namespace, w.changeMask, func(e changestream.ChangeEvent) bool {
> -		return e.Changed() == w.changeValue
> -	})
> +	opt := changestream.FilteredNamespace(w.namespace, w.changeMask, w.filter)
>  	subscription, err := w.watchableDB.Subscribe(opt)
>  	if err != nil {
>  		return errors.Annotatef(err, "subscribing to namespace %q", w.namespace)
> @@ -101,7 +140,7 @@ func (w *ValueWatcher) loop() error {
>  			return ErrSubscriptionClosed
>  		case changes, ok := <-in:
>  			if !ok {
> -				w.logger.Debugf("change channel closed for %q; terminating watcher for %q", w.namespace, w.changeValue)
> +				w.logger.Debugf("change channel closed for %q; terminating watcher", w.namespace)
>  				return nil
>  			}
>  
> @@ -129,7 +168,7 @@ func (w *ValueWatcher) drainInitialEvent(in <-chan []changestream.ChangeEvent) {
>  	select {
>  	case _, ok := <-in:
>  		if !ok {
> -			w.logger.Debugf("change channel closed for %q; terminating watcher for %q", w.namespace, w.changeValue)
> +			w.logger.Debugf("change channel closed for %q; terminating watcher", w.namespace)
>  			return
>  		}
>  	default:
> ```

Sure, good suggestion.

wallyworld commented on core/watcher/eventsource/value.go:
```diff
@@ -31,28 +31,42 @@ type ValueWatcher struct {
+	return NewValueMapperWatcher(base, namespace, changeValue, changeMask, defaultMapper)
+}
+
+// NewValueMapperWatcher returns a new watcher that receives changes from
+// the input base watcher's db/queue when mapper accepts the change-log
+// events occur for a specific changeValue from the input namespace.
```
Existing issue, but sentence doesn't quite parse.

wallyworld reviewed (approved):
Looks ok with the changes Simon suggested.
Make sure the jira card is a link.

ycliuhw commented:
/merge
